components/sup/src/manager/spec_dir.rs

```diff
@@ -0,0 +1,141 @@
+                            spec_file.display(),
+                            err
+                        );
+                    }
+                }
+                Err(err) => {
+                    outputln!(
+                        "Unable to migrate service spec, {}, {}",
+                        spec_file.display(),
+                        err
+                    );
+                }
+            }
+        }
+    }
+
+    /// Return a list of all the specs as currently found on disk.
+    pub fn specs(&self) -> Result<Vec<ServiceSpec>> {
+        let mut specs = vec![];
+
+        for spec_file in self.spec_files() {
+            let spec = match ServiceSpec::from_file(&spec_file) {
+                Ok(s) => s,
+                Err(e) => {
+                    match e.err {
+                        // If the error is related to loading a `ServiceSpec`, emit a warning
+                        // message and continue on to the next spec file. The best we can do to
+                        // fail-safe is report and skip.
+                        Error::ServiceSpecParse(_) | Error::MissingRequiredIdent => {
+                            outputln!(
+                                "Error when loading service spec file '{}' ({}). \
+                                 This file will be skipped.",
+                                spec_file.display(),
+                                e.description()
+                            );
+                            continue;
+                        }
+                        // All other errors are unexpected and should be dealt with up the calling
+                        // stack.
+
+                        // TODO (CM): This is the only way this
+                        // function could fail.
+                        _ => return Err(e),
+                    }
+                }
+            };
+
+            let file_stem = match spec_file.file_stem().and_then(OsStr::to_str) {
+                Some(s) => s,
+                None => {
+                    outputln!(
+                        "Error when loading service spec file '{}' \
+                         (File stem could not be determined). \
+                         This file will be skipped.",
+                        spec_file.display()
+                    );
+                    continue;
+                }
+            };
+
+            if file_stem != &spec.ident.name {
+                outputln!(
+                    "Error when loading service spec file '{}' \
+                     (File name does not match ident name '{}' from ident = \"{}\", \
+                     it should be called '{}.{}'). \
+                     This file will be skipped.",
+                    spec_file.display(),
+                    &spec.ident.name,
+                    &spec.ident,
+                    &spec.ident.name,
+                    SPEC_FILE_EXT
+                );
+                continue;
+            }
```

Comment:
Right now, the separate `let` (L100) and `if` (L113) blocks are logically coupled, but structurally independent. If the `continue` on L125 were accidentally removed, there would be a logic error. Instead, we can couple them together and make the code more robust:
```rust
            specs.push(match spec_file.file_stem().and_then(OsStr::to_str) {
                Some(stem) if stem == spec.ident.name => spec,
                Some(_) => {
                    outputln!( … );
                    continue;
                }
                None => {
                    outputln!( … );
                    continue;
                }
            });
```
Or if you don't like the `match` inside the `push` call, this version still has most of the same benefits:
```rust
            match spec_file.file_stem().and_then(OsStr::to_str) {
                Some(stem) if stem == spec.ident.name => spec,
                Some(_) => {
                    outputln!( … );
                    continue;
                }
                None => {
                    outputln!( … );
                    continue;
                }
            };

            specs.push(spec);
```
But it's a little less obvious why we're returning `spec` from the first arm.

Reply:
Excellent catch.